execute-async-function.js

```diff
@@ -52,27 +52,46 @@
         return execArgs;
     }
 
-    /** Create a promise that resolves when chrome.runtime.onMessage fires with the id
+    /** Create a promise that resolves when a chrome event (such as chrome.runtime.onMessage) fires
+     * @param {Event} chromeEvent The event type.
      * @param {string} id ID for the message we're expecting.
      * Messages without the ID will not resolve this promise.
      * @returns {Promise} Promise that resolves when chrome.runtime.onMessage.addListener fires. */
-    function promisifyRuntimeMessage(id) {
-        // We don't have a reject because the finally in the script wrapper should ensure this always gets called.
+    function promisifyChromeEvent(chromeEvent, id) {
+        // // We don't have a reject because the finally in the script wrapper should ensure this always gets called.
+        // We don't have a reject because this should be designed to always get called.
         return new Promise(resolve => {
-            const listener = request => {
-                // Check that the message sent is intended for this listener
-                if (request && request.asyncFuncID === id) {
-
-                    // Remove this listener
-                    chrome.runtime.onMessage.removeListener(listener);
-                    resolve(request);
+            const listener = function(...params) {
+                
+                if (chromeEvent === chrome.tabs.onUpdated) {
+                    const tabId = params[0];
+                    const changeInfo = params[1];
+                    const tab = params[2];
+                    // onUpdated event is called multiple times during a single load.
+                    // the status of 'complete' is called only once, when it is finished.
+                    if (tabId === id && changeInfo.status === 'complete') { 
+                        // Remove this listener
+                        chromeEvent.removeListener(listener);
+                        resolve({
+                            tabId: tabId,
+                            changeInfo: changeInfo,
+                            tab: tab
+                        });
+                    }
+                } else { // if (chromeEvent === 'chrome.runtime.onMessage')

```

Comment:
Where previously only `chrome.runtime.onMessage` was supported, now you can pass `chrome.tabs.onUpdated` and the new code runs, `chrome.runtime.onMessage` and the old code works, or any other event and get something unpredictable happening.